### PR TITLE
Double-line graph

### PR DESCRIPTION
**Pull request by jianhandev**
Screenshot 2021-01-29 at 5 07 04 PM

Other changes:
- moved time series graph to separate page

### CODE REVIEW

**Comment by moziliar** [comment]
Very Cool! Can you make another y axis to annotate one of the two lines? (I think the current y axis annotates the change)

File: src/main/java/teammates/storage/api/FeedbackResponseMonitorDb.java [hunk 1]
```diff
@@ -45,7 +45,7 @@ FeedbackResponseRecordAttributes makeAttributes(FeedbackResponseRecord entity) {
      * Gets a set of response records with chosen duration and interval.
      */
     public List<FeedbackResponseRecordAttributes> getResponseRecords(long duration, long interval) {
-        long startTimeInMilisec = System.currentTimeMillis() - duration;
+        long startTimeInMilisec = System.currentTimeMillis() - duration * 1000;
```
Comment (moziliar) [inline src/main/java/teammates/storage/api/FeedbackResponseMonitorDb.java]: Now that you have done it... Can frontend send in milliseconds instead? This looks very out of place.

File: src/main/java/teammates/ui/webapi/CreateAccountAction.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ private String importDemoData(String instructorEmail, String instructorName)
             }
 
             count -= rand.nextInt(numResponses / numIntervals);
-            timestamp -= 120; // 2 minutes interval
+            timestamp -= (30 * 1000); // 2 minutes interval
```
Comment (moziliar) [inline src/main/java/teammates/ui/webapi/CreateAccountAction.java]: Sin of having comment... If you are to change the value to mean something, please change the comment alongside... Although I should have made a named variable for this number which is now a magic number.

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.html [hunk 3]
```diff
@@ -5,11 +5,11 @@
     <div>
         <label>Duration:</label>
         <select class="form-control" [ngModel]="model.durationMinutes" (ngModelChange)="setDurationHandler($event)">
-            <option *ngFor="let duration of [1, 5, 10, 15, 30, 60]" [ngValue]="duration">{{ duration + ' min'}}</option>
+            <option *ngFor="let duration of [5, 10, 15, 30, 60]" [ngValue]="duration">{{ 'Past ' + duration + ' mins'}}</option>
```
Comment (moziliar) [inline src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: Can we have a more diverse options? Like 15min, 1hr, 3hr, 6hr, 24hr, 2days.
Seeing 5, 10, 15, 30, 60 feels a bit useless, as the graph does not change that drastically, and you can infer 30 minute graph from a 60 minute graph easily.

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.html [hunk 4]
```diff
@@ -5,11 +5,11 @@
...
         </select>
         <label>Interval:</label>
         <select class="form-control" [ngModel]="model.intervalSeconds" (ngModelChange)="setIntervalHandler($event)">
-            <option *ngFor="let interval of [10, 15, 30, 60]" [ngValue]="interval">{{ interval + ' sec'}}</option>
+            <option *ngFor="let interval of [10, 15, 30, 60]" [ngValue]="interval">{{ interval + ' secs'}}</option>
```
Comment (moziliar) [inline src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: "refresh interval"?

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts [hunk 5]
```diff
@@ -40,10 +42,29 @@ export class ResponseTimeSeriesChartComponent implements OnInit {
...
-          console.log(records);
-          this.drawChart(records.responseRecords, durationSeconds * 1000);
+          const canvas: any = d3.select('svg');
+
+          // clear all content
+          canvas.selectAll('*').remove();
+
+          this.drawChart(canvas, records.responseRecords, durationSeconds * 1000, false, '#696969');
+
+          const temp: number[] = [];
```
Comment (moziliar) [inline src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts]: Can this be named more meaningfully? e.g. totalCounts

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts [hunk 6]
```diff
@@ -40,10 +42,29 @@ export class ResponseTimeSeriesChartComponent implements OnInit {
...
+          records.responseRecords.forEach((record: FeedbackResponseRecord, i: number) => {
+            temp[i] = record.count;
+
+            if (i <= 0) {
+              record.count = 0;
+            } else {
+              record.count = record.count - temp[i - 1];
+            }
+
+            return record;
```
Comment (moziliar) [inline src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts:65]: forEach takes in a Consumer... I believe in 2030 we have learnt what a Consumer does. What does the return here mean?

**Comment by jianhandev** [comment]
> Very Cool! Can you make another y axis to annotate one of the two lines? (I think the current y axis annotates the change)

A y-axis on the right hand side right, I was thinking about that too

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts [hunk 7]
```diff
@@ -40,10 +42,29 @@ export class ResponseTimeSeriesChartComponent implements OnInit {
...
-          this.drawChart(records.responseRecords, durationSeconds * 1000);
+          const canvas: any = d3.select('svg');
+
+          // clear all content
+          canvas.selectAll('*').remove();
+
+          this.drawChart(canvas, records.responseRecords, durationSeconds * 1000, false, '#696969');
+
+          const temp: number[] = [];
+          records.responseRecords.forEach((record: FeedbackResponseRecord, i: number) => {
```
Comment (moziliar) [inline src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts]: This entire thing seems very odd...
Could read better if you use a for loop instead.
I believe you mean `map` here but found forEach instead.

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts [hunk 8]
```diff
@@ -62,21 +84,19 @@ export class ResponseTimeSeriesChartComponent implements OnInit {
     this.runInterval.next(newInterval);
   }
 
-  drawChart(data: FeedbackResponseRecord[], duration: number): void {
-    const svg: any = d3.select('svg');
+  drawChart(canvas: any, data: FeedbackResponseRecord[], duration: number,
```
Comment (moziliar) [inline src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts]: Justify your use of `any` type here.

**Review by moziliar — CHANGES_REQUESTED** [review]

**Comment by moziliar** [comment]
@jianhandev 
Yes, otherwise the graph looks extremely confusing (what the y axis indicates? the rising graph or the fluctuating one? what do the data points mean?)

File: src/main/java/teammates/storage/api/FeedbackResponseMonitorDb.java (hunk 1 above)
Comment (daongochieu2810) [inline src/main/java/teammates/storage/api/FeedbackResponseMonitorDb.java]: I thought the front end would send milliseconds, why need to multiply?

**Comment by daongochieu2810** [comment]
Add padding also please :)))

File: src/main/java/teammates/storage/api/FeedbackResponseMonitorDb.java (hunk 1 above)
Comment (jianhandev) [reply src/main/java/teammates/storage/api/FeedbackResponseMonitorDb.java]: will change to milliseconds, was in seconds

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.html (hunk 3 above)
Comment (jianhandev) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: How about 5min, 30min, 1 hour, 12 hours, 1 day, 1 week, 1 month?

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.html (hunk 4 above)
Comment (jianhandev) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: Screenshot 2021-01-28 at 7 30 18 PM

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts (hunk 7 above)
Comment (jianhandev) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts]: forEach seems to work too, in javascript, but yeah i agree `map` would be more apt

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts (hunk 8 above)
Comment (jianhandev) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts]: `const canvas: any = d3.select('svg');` i cant find a type for this

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.html (hunk 3 above)
Comment (daongochieu2810) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: ^ I think this works

Comment (moziliar) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: ditto, i think the scale works well

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts (hunk 8 above)
Comment (moziliar) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.ts]: Can't verify as I don't have the permission to pull your code.
Let me see after we merge this

File: src/web/app/components/response-time-series-chart/response-time-series-chart.component.html (hunk 3 above)
Comment (moziliar) [reply src/web/app/components/response-time-series-chart/response-time-series-chart.component.html]: yes

**Comment by moziliar** [comment]
- Spacing between the interval and duration, too cramped

- The time display can show in 12h? A bit confusing looking at it now. If possible can let user choose? This is just an add-on, I think in general the target is achieved

UPDATE: I think the x axis is fine if you delete the :30 ones, coz they are easily inferable and is quite distracting

**Pull request merged**